### PR TITLE
ISPN-10946 Tests should use a MetricsRegistry per cache manager insta…

### PR DESCRIPTION
**Pull request by anistor**
…nce rather than per JVM

https://issues.jboss.org/browse/ISPN-10946

### CODE REVIEW

**Comment by pruivo** [comment]
@anistor there are a lot of failures in CI

**Comment by anistor** [comment]
That's why it is called CI @pruivo ! So I don't have to smoke test them here melting this laptop :))

**Comment by anistor** [comment]
Rebased, and hopefully failures are fixed also. Let's wait for CI

File: core/src/main/java/org/infinispan/metrics/impl/ApplicationMetricsRegistry.java [hunk 1]
```diff
@@ -34,19 +38,33 @@
...
       return applicationMetricRegistry;
    }
 
    public void register(ResourceDMBean resourceDMBean) {
+      MetricRegistry registry = getRegistry();
+
       ObjectName objectName = resourceDMBean.getObjectName();
+      int metricCounter = 0;
+      if (log.isDebugEnabled()) {
+         log.debugf("Metric registry @%x contains %d metrics. Registering metrics for ObjectName \"%s\" ...",
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/metrics/impl/ApplicationMetricsRegistry.java]: IMO this is too low-level even for debug, it and the other messages you added should be at the trace level.

File: core/src/test/java/org/infinispan/test/fwk/TestApplicationMetricsRegistry.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ */
+final class TestApplicationMetricsRegistry extends ApplicationMetricsRegistry {
+
+   @Override
+   protected MetricRegistry makeRegistry() {
+      return new MetricsRegistryImpl();
+   }
+
+   static void replaceMetricsRegistry(DefaultCacheManager cacheManager) {
+      TestingUtil.replaceComponent(cacheManager, ApplicationMetricsRegistry.class, new TestApplicationMetricsRegistry(), true);
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/test/fwk/TestApplicationMetricsRegistry.java]: Aren't some metrics already registered by the time you replace the component here?

I was really hoping you'd fix `GlobalConfigurationBuilder.addModule()` to not replace existing modules and to use `global.addModule(TestGlobalConfigurationBuilder.class).testGlobalComponent(ApplicationMetricsRegistry.class, new TestApplicationMetricsRegistry())` instead.

File: core/src/test/java/org/infinispan/test/fwk/TestCacheManagerFactory.java [hunk 3]
```diff
@@ -326,13 +328,9 @@ public static EmbeddedCacheManager createCacheManager(SerializationContextInitia
...
       ConfigurationBuilder builder = new ConfigurationBuilder();
-      builder
-            .clustering()
-            .cacheMode(mode)
-            .indexing()
-            .index(indexing ? Index.ALL : Index.NONE)
-            .addProperty("lucene_version", "LUCENE_CURRENT")
-      ;
+      builder.clustering().cacheMode(mode)
+            .indexing().index(indexing ? Index.ALL : Index.NONE)
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/test/fwk/TestCacheManagerFactory.java]: Nitpicking: Please enable "align when multiline" in IntelliJ for chained method calls, method call arguments, and method declaration parameters, I've gotten used to it and I like it a lot more than fixed indents nowadays.

Comment (anistor) [reply core/src/test/java/org/infinispan/test/fwk/TestCacheManagerFactory.java]: no way :)

File: core/src/test/java/org/infinispan/test/fwk/TestApplicationMetricsRegistry.java (hunk 2 above)
Comment (anistor) [reply core/src/test/java/org/infinispan/test/fwk/TestApplicationMetricsRegistry.java]: done

File: core/src/main/java/org/infinispan/metrics/impl/ApplicationMetricsRegistry.java (hunk 1 above)
Comment (anistor) [reply core/src/main/java/org/infinispan/metrics/impl/ApplicationMetricsRegistry.java]: done

**Comment by anistor** [comment]
Just updated this @danberindei . I'd like it merged in beta1.  My local run indicates the failures are gone, but let's wait for CI.

File: core/src/test/java/org/infinispan/commands/module/TestModuleLifecycle.java [hunk 4]
```diff
@@ -59,7 +58,7 @@ public void cacheManagerStarting(GlobalComponentRegistry gcr, GlobalConfiguratio
 
       Set<String> componentNames = testGlobalConfiguration.globalTestComponents().keySet();
       for (String componentName : componentNames) {
-         AssertJUnit.assertSame(testGlobalConfiguration.globalTestComponents().get(componentName), gcr.getComponent(componentName));
+         assert testGlobalConfiguration.globalTestComponents().get(componentName) == gcr.getComponent(componentName);
```
Comment (anistor) [inline core/src/test/java/org/infinispan/commands/module/TestModuleLifecycle.java]: This should not depend on the test framework or will needlessly propagate testng dependency in pure junit tests.

Comment (danberindei) [reply core/src/test/java/org/infinispan/commands/module/TestModuleLifecycle.java]: There's no such thing as a pure JUnit test depending on core test-jar :)

File: core/src/main/java/org/infinispan/metrics/impl/ApplicationMetricsRegistry.java [hunk 5]
```diff
@@ -78,13 +96,40 @@ public void register(ResourceDMBean resourceDMBean) {
...
+         log.tracef("Metric registry @%x contains %d metrics. Unregistering metrics for ObjectName \"%s\" ...",
+                    System.identityHashCode(registry), registry.getMetrics().size(), objectName);
+      }
       String prefix = ObjectNameMapper.makeMetricNamePrefix(objectName);
       Map<String, String> tags = ObjectNameMapper.makeTagMap(objectName);
-      getRegistry().removeMatching((metricID, metric) -> metricID.getName().startsWith(prefix) && tags.equals(metricID.getTags()));
+      AtomicInteger metricCounter = new AtomicInteger();
+      registry.removeMatching((metricID, metric) -> {
+         boolean isMatching = metricID.getName().startsWith(prefix) && tags.equals(metricID.getTags());
+         if (log.isDebugEnabled() && isMatching) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/metrics/impl/ApplicationMetricsRegistry.java]:
```suggestion
         if (log.isTraceEnabled() && isMatching) {
```

**Comment by danberindei** [comment]
Integrated, thanks Adrian!

File: core/src/test/java/org/infinispan/commands/module/TestModuleLifecycle.java (hunk 4 above)
Comment (anistor) [reply core/src/test/java/org/infinispan/commands/module/TestModuleLifecycle.java]: in that case, bleahhhhhh :nauseated_face:

**Pull request closed without merging**